clang.cpp

```diff
@@ -1,5 +1,37 @@
-// This file includes functions needed to build with clang.
+// This file includes functions required to build VP using Clang.
+
+// Intrinsics required by code generated using VS2022 `LINK /cvtcil`.
 extern "C" {
-	long _ftol3(float f) { return static_cast<long>(f); }
-	double _ltod3(long l) { return static_cast<double>(l); }
+    /// Convert a single precision floating point input as `xmm0` to a
+    /// int64 with the low 32-bits in `eax` and high 32-bits in `edx`.
+    __attribute__((naked)) void _ftol3() {
+        // Clang 14 generated SSE3 dependent implementation.
+        __asm__(".intel_syntax\n"
+            "sub     esp, 12\n"
+            "movss   dword ptr[esp], xmm0\n"
+            "fld     dword ptr[esp]\n"
+            "fisttp  qword ptr[esp]\n"
+            "mov     eax, dword ptr[esp]\n"
+            "mov     edx, dword ptr[esp + 4]\n"
+            "add     esp, 12\n"
+            "ret\n"
+        );
+    }
+    /// Convert an int64 with the low 32-bits in `ecx` and the high 32-bits
+    /// in `edx` to a double precision floating point output as `xmm0`.
+    __attribute__((naked)) void _ltod3() {
+        // Clang 14 generated SSE3 dependent implementation.
+        __asm__(".intel_syntax\n"
+            "sub     esp, 20\n"
+            "movd    xmm0, edx\n"
+            "movd    xmm1, ecx\n"
```

Comment:
are you sure _ltod3 is allowed to clobber xmm1? also, are you sure this can't be fixed by linking against msvcrt or somesuch?

Reply:
These functions are normally provided by the VS2022 CRT. I'm not sure that linking against multiple CRTs is feasible. As for clobbering `xmm1`, Microsoft's implementation already does this along the branch where AVX instructions are unavailable so it's presumably safe to do so.